### ISSUE TITLE
metadata: Add payload length to marshalled binary

### ISSUE THREAD

**Issue by MarcoPolo** [ipni/go-libipni#27]
Here's my understanding of how metadata is marshalled:

1. Sort all the protocols.
2. Ask each protocol to marshal itself, and append this to our output buffer.
3. Return the output buffer.

Bitswap will write its multicodec.
GraphsyncFilecoinV1 will write its multicodec followed by some dag-cbor bytes.

Note that we don't write that payload length anywhere here (bitswap would be 0, GraphsyncFilecoinV1 would be the length of the dag-cbor bytes). This becomes a problem for consumers of the bytes, since if they don't know about a protocol, they have no way of "skipping" it. Imagine we have a protocol with Code=1, without knowing about what that protocol writes as its payload you won't be able to interpret the rest of the payload. This protocol could have length=0 or length=totalbuffer size.

My ask is that from here on out (with the possible exception of bitswap/graphsync for backwards compat) we always marshall things as `<codec><payload-length><payload>`, this will allow consumers to skip over the codecs they don't know about or don't care for.

I _think_ this is the idea behind the `metadata.Unknown`, except I don't think that currently works right now. It tries to read the payload length in `ReadFrom`, but nothing actually writes the payload length.

**Comment by masih** [comment]
> always marshall things as <codec><payload-length><payload>

Agreed. this is true for HTTP gateway codec recently added. Happy to add this to IPNI specification; PRs are welcome.
